lib/events/filelog.go

```diff
@@ -136,12 +156,30 @@ func (l *FileLog) EmitAuditEvent(ctx context.Context, event AuditEvent) error {
 
 // EmitAuditEventLegacy adds a new event to the log. Part of auth.IFileLog interface.
 func (l *FileLog) EmitAuditEventLegacy(event Event, fields EventFields) error {
+	l.rw.RLock()
+	defer l.rw.RUnlock()
+
 	// see if the log needs to be rotated
-	err := l.rotateLog()
-	if err != nil {
-		log.Error(err)
+	if l.mightNeedRotation() {
+		// log might need rotation; switch to write-lock
+		// to avoid rotating during concurrent event emission.
+		l.rw.RUnlock()
+		l.rw.Lock()
+
+		// perform rotation if still necessary (rotateLog rechecks the
+		// requirements internally, since rotation may have been performed
+		// during our switch from read to write locks)
+		err := l.rotateLog()
```

Comment:
Nit: Lock won't be released if this call panics but don't know if that's a big risk. Maybe moving lock back inside rotateLog would help cause you could defer unlock then.

Reply:
Started with the locking inside `rotateLog` but I decided that was better to perform it inside of the outer scope in order to ensure that future updates to calling functions do not accidentally assume lock continuity.  IMO if we start using panics as control flow, we're going to have to rewrite a ton of our code anyway, so I'd rather stick with the layout that minimizes developer error under normal circumstances for now.